Rearrange report generation in qualification Jenkinsfile

james-smith-za opened the pull request:
There was a try/finally block. The only thing that was in 'try' was the pytest run itself. But if there's a problem with the report generation, this doesn't get caught.

So I moved the execution of script to generate the report (and in fact the procedure document) into the `try` block, added a catch just to publish the json if everything else goes wrong, and then the finally remains basically as it was for whatever is left.

I have no idea if this will work, but I'm unable to retrieve the JSON from failed Jenkins runs so this is my attempt at debugging that. I can't reproduce the `ExceptionChainRepr` thing so far with just running qualification tests manually from the CLI.

Contributes to NGC-875

Checklist (if not applicable, edit to add `(N/A)` and mark as done):

- [x] (N/A) If dependencies are added/removed: update `setup.cfg` and `.pre-commit-config.yaml`
- [x] (N/A) If modules are added/removed: use sphinx-apidoc to update files in `doc/`
- [x] Ensure copyright notices are present and up-to-date
- [x] (N/A) If qualification tests are changed: attach a sample qualification report
- [x] (N/A) If design has changed: ensure documentation is up to date
- [x] (N/A) If ICD-defined sensors have been added: update `fake_servers.py` in katsdpcontroller to match

ijassiem reviewed (approved):
We can ran a test on this branch before merging.

james-smith-za commented:
Thanks Ismail.

Would you be able to kick that off for me please?

On Thu, Feb 23, 2023 at 9:00 AM Ismail ***@***.***> wrote:

> @ijassiem approved this pull request.We can ran a test on this branch
> before merging.—Reply to this email directly, view it on GitHub, or
> unsubscribe.You are receiving this because you authored the thread.Message
> ID: ***@***.***
>  ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌
>
> <https://za.report.cybergraph.mimecast.com/alert-details/?dep=t%2BZp07ZndZKcOWu%2Fdqg4cQ%3D%3D%2FZYu4jCkBtEI5o1MontYdQ5SLvKazB5spcXuFYIOXfaMDdgwOJJ7hs9mOB01kGYRgoM5kFYMFJGyP94lwXMG12PuMBV2nLrbqWFX2VfWca9bwrpKiU6lxnp53zY8WDgM%2BviSiiR17nmwBdMSie%2FhiaW8jBPh8uiYxgZ1pz%2BHqvow2CfyLZ0wgpOcuysrRFEjyTJ2K%2BB4XyPrKdn376Fx7DlC3MfkEz2rdtU6yBktJANdstpPJ0NYd%2B4kYq7LAWE3rtvDehpqB4%2Bj8FYY9GMA8YY8Aq1Ty%2BrH1qrKMSbhXZ%2Fm9qpvqVlzykJ5HHyoK9nRWhh1aQ6a95Z5zr0dIQCypW0DznI%2Bu39thnREKfcI5cjaVYJOeDBiB91fVDwCuYKizbJ8ei1VPIg7%2FWw8zpwnhv0QLi7m16wsfo3eVIE65qDhWI41NYA%2BZRhnHTScpO9DsHhh%2B5Lfy88nPjcFsTHAUTSHvAaXvrq%2FxhyP4a8%2BL9yObzPshOtpAgurQNj0trZsDSeEkUVHbvQHqw0rUVCaEseXBDNknr3zyUIFFvR42A%2BeE5wX1xpSYQP2vWaZsCGla22kL3PA5kNXSsF11iyrvxQytGAhbKph0jkSMnGvY0YTvkYZSsD3upkMzNBLDxORAtwz%2B171bEtafTWwgNpa3Z%2Fm9kcm%2FQ5f%2BGeXHoBIVZ4PGCUQBqWteRWqWloGHJIugzifbwiC0sRqUFNkdESkNA%3D%3D>
>
> ***@***.**** approved this pull request.
>
> We can ran a test on this branch before merging.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ska-sa/katgpucbf/pull/526#pullrequestreview-1310696065>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACS4QM4PNWFIQOZHJNBQ7BTWY4DJFANCNFSM6AAAAAAVEKLSHM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Disclaimer

The information contained in this communication from the sender is confidential. It is intended solely for use by the recipient and others authorized to receive it. If you are not the recipient, you are hereby notified that any disclosure, copying, distribution or taking action in relation of the contents of this information is strictly prohibited and may be unlawful.

This email has been scanned for viruses and malware, and may have been automatically archived by Mimecast, a leader in email security and cyber resilience. Mimecast integrates email defenses with brand protection, security awareness training, web security, compliance and other essential capabilities. Mimecast helps protect large and small organizations from malicious activity, human error and technology failure; and to lead the movement toward building a more resilient world. To find out more, visit our website.

amishatishpatel commented:
Jenkins has completed, and I think it's succeeded (according to the screenshot below). Happy to merge if you are @ijassiem.
![image](https://user-images.githubusercontent.com/25218138/220859478-d0be87c7-af49-4e2d-990f-dd2a94528a77.png)

ijassiem commented:
Yes looks all good.

james-smith-za commented:
I wasn't entirely happy with this outcome, because Jenkins was reporting "good" when there were tests failing. I think this was due to the way that I'd handled the catch block. I've now nested this and tried to re-throw the exception if report-construction breaks. This should alert us if stuff goes wrong, but also allow us to get the intermediate product out and / or the report.

@ijassiem if you could kick off another build on this branch then I'd appreciate it.

james-smith-za commented:
Ok up to now, even with my latest changes, it still reports "SUCCESS" when the tests have in fact failed.

So there must be something that I'm missing.

ijassiem commented on qualification/Jenkinsfile:
```diff
@@ -73,18 +73,28 @@ pipeline {
-             publishHTML(target: [keepAll: true, reportName: 'Qualification Test Procedure', reportDir: '', reportFiles: 'procedure.pdf', reportTitles: 'procedure.pdf'])
-             currentBuild.displayName = "#${BUILD_NUMBER} (${COMMIT_ID})"
-           }
+           sh 'spead2_net_raw pytest -v -c qualification/pytest-jenkins.ini qualification --image-override katgpucbf:harbor.sdp.kat.ac.za/cbf/katgpucbf:latest --junitxml=result.xml'
```
Can we have this line of code in a try block to catch errors with the starting up of a correlator, or bugs in the running of the qualification tests?

james-smith-za replied:
The issue that I have is that, unless there's some Jenkins magic that I don't understand, you catch all problems, whether it's a test that works but doesn't meet spec, or an actual bug in the process somehow.

Currently this is what's happening:
![image](https://user-images.githubusercontent.com/10864691/221161773-a872ed05-ede1-4862-bb1e-af17d3196d85.png)

which is exactly what I want. The tests are failing, so it lets us know about it, and the report generation isn't working, so that block is failing as well.

@amishatishpatel would you be able to weigh in?

james-smith-za commented:
Though now that I'm looking at the Blue Ocean output, it looks as though it's not even getting to the "publish" stage:
![image](https://user-images.githubusercontent.com/10864691/221165293-f0f70adc-3186-470e-8cec-0ce5f95d4183.png)

james-smith-za commented:
@bmerry I think that I'm more or less happy with the pipeline stages now.
The only thing I'm not quite certain about is what the json is supposed to look like when I click on it. Currently I just see a white page which takes about a minute and a half to load up. I don't know if Jenkins is transparently decompressing things.
I'm not sure where on the filesystem to go and inspect what's actually going on. I may be wildly misunderstanding the inner workings.

bmerry commented:
I'm not wild about moving so much logic into the `post` action. I think that should be just about cleaning up and ensuring an email get sent. At the moment if something goes wrong in the post action, it's going to bail out of it, and the email won't be generated.

I think a better way to get the behaviour you're aiming for (of producing as much reporting as possible when something goes wrong) is to interleave the publishing steps with the processing steps. So for example, publish the JSON before trying to generate the report, then publish the report before trying to generate the procedure etc.

I still need to figure out how the return code of 'sh' steps interacts with failures - I thought I understood it, but the main Jenkinsfile is having me question myself.

bmerry commented:
> I still need to figure out how the return code of 'sh' steps interacts with failures - I thought I understood it, but the main Jenkinsfile is having me question myself.

Ok, it turns out that when unit tests fail, we don't publish the unit tests results. That's a separate bug we should look at.

Here's a suggestion:
1. Use something like https://github.com/yashtodi94/pytest-custom_exit_code, or just post-process the exit code, so that the Jenkins step doesn't fail overall if some tests fail.
2. Rely on the 'junit' step to mark the build as "unstable" if there were failed tests.

IIRC, "unstable" doesn't prevent further steps from occurring, so we should then still get the report and procedure etc, without the need for any try/finally magic.

bmerry commented on qualification/Jenkinsfile:
```diff
@@ -73,20 +73,33 @@ pipeline {
+           publishHTML(target: [keepAll: true, reportName: 'Qualification Test Report', reportDir: '', reportFiles: 'report.pdf', reportTitles: 'report.json'])
+
+           // If that worked, we can probably generate and publish a procedure as well
+           sh(script: 'qualification/report/generate_pdf.py report.json procedure.pdf --generate-procedure-doc')
```
```suggestion
           sh 'qualification/report/generate_pdf.py report.json procedure.pdf --generate-procedure-doc'
```
For consistency with most other uses of `sh`.

bmerry commented on qualification/Jenkinsfile:
```diff
@@ -73,20 +73,33 @@ pipeline {
+
+     stage('Publish results') {
+       steps {
+         script {
```
Most of this block doesn't need to be wrapped in `script`. It's for when you need an escape hatch to write general Groovy code instead of just defining steps declaratively. I *think* the only bits that need to be wrapped in `script` are the two lines involving COMMIT_ID. See https://www.jenkins.io/doc/book/pipeline/syntax/#script.

bmerry commented on qualification/Jenkinsfile:
```diff
@@ -73,20 +73,33 @@ pipeline {
+         script {
+           // Compress and publish json file so that at least we have that.
+           sh 'xz --keep --force -T 0 report.json'
+           publishHTML(target: [keepAll: true, reportName: 'Qualification Test Intermediate JSON', reportDir: '', reportFiles: 'report.json.xz', reportTitles: 'report.json.xz'])
```
Can remove `reportTitles`: if omitted, the filename is used as the title (https://www.jenkins.io/doc/pipeline/steps/htmlpublisher/#publishhtml-publish-html-reports).
```suggestion
           publishHTML(target: [keepAll: true, reportName: 'Qualification Test Intermediate JSON', reportDir: '', reportFiles: 'report.json.xz'])
```

bmerry commented on qualification/Jenkinsfile:
```diff
@@ -73,20 +73,33 @@ pipeline {
+           // Generate and publish test report
+           COMMIT_ID = sh(script: 'qualification/report/generate_pdf.py report.json report.pdf -c', returnStdout: true)
+           currentBuild.displayName = "#${BUILD_NUMBER} (${COMMIT_ID})"
+           publishHTML(target: [keepAll: true, reportName: 'Qualification Test Report', reportDir: '', reportFiles: 'report.pdf', reportTitles: 'report.json'])
```
The reportTitles is wrong (and unnecessary)
```suggestion
           publishHTML(target: [keepAll: true, reportName: 'Qualification Test Report', reportDir: '', reportFiles: 'report.pdf'])
```

bmerry commented on qualification/Jenkinsfile:
```diff
@@ -73,20 +73,33 @@ pipeline {
          script {
```
Since we're not writing general Groovy code (try/finally) we don't need to wrap these steps in a `script` block any more.

bmerry commented on qualification/Jenkinsfile:
```diff
@@ -73,20 +73,33 @@ pipeline {
-             publishHTML(target: [keepAll: true, reportName: 'Qualification Test Procedure', reportDir: '', reportFiles: 'procedure.pdf', reportTitles: 'procedure.pdf'])
-             currentBuild.displayName = "#${BUILD_NUMBER} (${COMMIT_ID})"
-           }
+           sh 'spead2_net_raw pytest -v -c qualification/pytest-jenkins.ini qualification --suppress-tests-failed-exit-code --image-override katgpucbf:harbor.sdp.kat.ac.za/cbf/katgpucbf:latest --junitxml=result.xml'
```
It's probably worth a comment to explain why we're using `--suppress-tests-failed-exit-code`.

james-smith-za commented on qualification/Jenkinsfile:
```diff
@@ -73,20 +73,33 @@ pipeline {
+           publishHTML(target: [keepAll: true, reportName: 'Qualification Test Procedure', reportDir: '', reportFiles: 'procedure.pdf', reportTitles: 'procedure.pdf'])
+
+           // Since everything is generated, we don't need the original anymore
+           sh 'rm -f report.json'
```
We were getting the error of `Already exists` previously on this step, which suggested to me that the cleanup wasn't actually happening.

bmerry replied:
I'm guessing that was because you had stuff in `post` that was failing, causing the rest of the `post` block to fail.

From the documentation it looks like you can have an additional `cleanup` block that follows `post`.

james-smith-za replied:
Ok, I just pushed a change removing this. I'm going to wait for the qual test to run, and then try and examine what's actually going on inside the Jenkins container to see whether it did clean up.

james-smith-za commented:
@bmerry I think I've addressed all your points.

I'm kicking off one more run, that should confirm whether everything is working, then if it passes (and I get approved) then I'm going to squash this PR into one commit.

bmerry commented on qualification/Jenkinsfile:
```diff
@@ -63,30 +63,56 @@ pipeline {
     stage('Run qualification tests') {
        steps {
+         /* Run the actual pytest.
+          * --supprress-tests-failed-exit-code argument does what it says, in
```
```suggestion
          * --suppress-tests-failed-exit-code argument does what it says, in
```

bmerry commented on qualification/Jenkinsfile:
```diff
@@ -63,30 +63,56 @@ pipeline {
+         /* Run the actual pytest.
+          * --supprress-tests-failed-exit-code argument does what it says, in
+          * case of failed tests, a nonzero exit code is produced, but that
+          * disrupts the Jenkins pipeline including generation of reports which
```
This is a run-on sentence. How about something like:

> Jenkins will not run subsequent steps if pytest returns a non-zero exit code. Passing --suppress-tests-failed-exit-code causes the exit code to be zero if there are failed tests, so that we can still get the report in that case. The junit step will detect the failed tests and mark the build as unstable.

bmerry commented on qualification/Jenkinsfile:
```diff
@@ -63,30 +63,56 @@ pipeline {
+
+     stage('Publish results') {
+       steps {
+         junit testResults: 'result.xml', skipPublishingChecks: true
```
Maybe worth a note to explain why skipPublishingChecks.

bmerry commented on qualification/Jenkinsfile:
```diff
@@ -118,6 +144,8 @@ pipeline {
           sh 'katcpcmd lab5.sdp.kat.ac.za:5001 product-deconfigure jenkins_qualification_correlator 1'
         }
       }
+    }
```
The last sentence of the comment above `post` should probably move down here.
